### PR TITLE
Implement telemetry for commands

### LINKED ISSUE #68: Track command usage with telemetry
We have download stats but we need a bit more details to help making decision for feature development.

We will introduce `kedro viz` command within the extension, we can add telemetry to get some insight about the usage. From my understanding, viz itself tracks the UI click but it wouldn't be sufficient to tell if it's from VSCode.

## Questions
Do we implement the telemetry in TS or Python? This may depends on what do we want to track.

- How many people are using the command  (and how often)
- Can we differentiate clicks from a browser versus inside vscode?

### Python
Pro:
- Leverage `kedro-telemetry` for consent
- Use heap to send relevant event as event triggered as a "command"
- Click event is embedded in flowchart (but is there a way to tell it's from VSCode?)

Con:
- Limited as not everything trigger LSP server (e.g. "restart environment" command)

### TS
Pro:
- Most flexible

Con:
- Need to implement telemetry from scratch maybe?

### PR DESCRIPTION
**Pull request by noklam**
# Context
Fixed #68 

# QA Notes:
- Set env variable so that the event send to dev server instead of prod, i.e. `HEAP_APP_ID = '4039408868'` (Before you launch VSCode, or export the env var and start vscode with `code .`
- Run any `Kedro` related command
- Inspect `Live Data feed` from https://heapanalytics.com/app/env/4039408868/data-galaxy?view=live-data-feed

image

# Dev Notes
- Reuse `kedro-telemetry` for the consent
- Run script as a tool to get consent / reuse LSP (python process) to get information about the consent
- Store the consent in extension.
- When `Kedro: run viz` or any command is triggered, send an event to HEAP using the server/client API.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -126,19 +137,28 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
...
             }
+            await sendHeapEventWithMetadata(CMD_SELECT_ENV);
         }),
         registerCommand('pygls.server.executeCommand', async () => {
             await executeServerCommand(lsClient);
         }),
-        registerCommand('kedro.sendDefinitionRequest', async (word) => {
+        registerCommand(CMD_DEFINITION_REQUEST, async (word) => {
             await executeServerDefinitionCommand(lsClient, word);
+            await sendHeapEventWithMetadata(CMD_DEFINITION_REQUEST);
```
Comment (noklam) [inline src/extension.ts:155]: Maybe we don't need to await this? @jitu5

File: bundled/tool/install_telemetry_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    libs_path = Path(extension_root_dir) / "bundled" / "libs"
+    requirements_path = Path(extension_root_dir) / "kedro-viz-requirements.txt"
+
+    try:
+        import kedro_telemetry
+        from packaging.version import parse
+
+        version = parse(kedro_telemetry.__version__)
+        if version.major<1 and version.minor<6: # at least >0.6.0
+            raise ImportError("kedro-telemetry version must be >=0.6.0")
```
Comment (jitu5) [inline bundled/tool/install_telemetry_dependencies.py:24]: Do we need to do version checks for kedro-viz dependencies as well ?

File: bundled/tool/install_viz_dependencies.py [hunk 3]
```diff
@@ -18,6 +18,9 @@ def install_dependencies(extension_root_dir):
     try:
         import fastapi
         import orjson
+        import kedro_telemetry
+        from packaging.version import parse
```
Comment (jitu5) [inline bundled/tool/install_viz_dependencies.py:22]: We can remove this right? we are already doing it in `bundled/tool/install_telemetry_dependencies.py`

File: src/common/telemetry.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import axios from 'axios';
```
Comment (jitu5) [inline src/common/telemetry.ts:1]: Do we have `axios` as dependency in package.json ?

File: src/common/utilities.ts [hunk 5]
```diff
@@ -71,22 +71,76 @@ export async function getProjectRoot(): Promise<WorkspaceFolder> {
...
-                console.log('Python dependencies installed!');
+            if (stdoutViz.includes('Successfully installed')) {
+                traceLog(`Kedro-viz dependencies installed!`);
+                console.log('Kedro-viz dependencies installed!');
+            }
+            if (stdoutTelemetry.includes('Successfully installed')) {
+                traceLog(`kedro-telemetry dependencies installed!`);
+                console.log('kedro-telemetry dependencies installed!');
             }
+            context.globalState.update(DEPENDENCIES_INSTALLED, true);
```
Comment (jitu5) [inline src/common/utilities.ts:92]: I think we need to set `DEPENDENCIES_INSTALLED` flag as `true` inside `stdoutViz.includes('Successfully installed')` and `stdoutTelemetry.includes('Successfully installed')`

File: src/common/utilities.ts [hunk 6]
```diff
@@ -71,22 +71,76 @@ export async function getProjectRoot(): Promise<WorkspaceFolder> {
...
+        // Check if the script output contains the success message
+        if (telemetryResult) {
+            const consent = telemetryResult['consent']
+            // Step 2: Create a Map from the record
+            const projectMetadata = new Map(Object.entries(telemetryResult));
+            context.globalState.update(PROJECT_METADATA, projectMetadata);
+            projectMetadata.delete('consent');
+
+            context.globalState.update(TELEMETRY_CONSENT, consent);
+            console.log(`Consent from Kedro Project: true ${consent}`);
```
Comment (jitu5) [inline src/common/utilities.ts:118]: We need to remove `true` I think.

File: src/extension.ts [hunk 7]
```diff
@@ -109,12 +108,24 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
...
+    let heapUserId: string = '';
+    projectMetadata = context.globalState.get(PROJECT_METADATA);
+    if (projectMetadata) {
+        heapUserId = projectMetadata.get("username");
+    }
+
+    const sendHeapEventWithMetadata = async (eventName: string): Promise<void> => {
+        sendHeapEvent(eventName, projectMetadata, heapUserId);
+
+    }
```
Comment (jitu5) [inline src/extension.ts:127]: Can we move this code to `/common/telemetry` same file as `sendHeapEvent` ?

File: src/extension.ts (hunk 1 above)
Comment (jitu5) [reply src/extension.ts:155]: Yes.

File: src/common/telemetry.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+// from kedro_telemetry.plugin
+interface HeapData {
+    app_id: string;
```
Comment (jitu5) [inline src/common/telemetry.ts:5]: [nit]:
```suggestion
    appId: string;
```

File: bundled/tool/install_telemetry_dependencies.py (hunk 2 above)
Comment (noklam) [reply bundled/tool/install_telemetry_dependencies.py:24]: @jitu5 I think so, can you address this separately in other PR?

File: bundled/tool/install_viz_dependencies.py (hunk 3 above)
Comment (noklam) [reply bundled/tool/install_viz_dependencies.py:22]: indeed - removed

File: src/common/telemetry.ts (hunk 4 above)
Comment (noklam) [reply src/common/telemetry.ts:1]: I have updated the `package.json`, I need something simple to send a HTTP request so `axios` is not a must here. If you know any libraries that can do the same job in existing dependencies, happy to remove `axios` too.

File: src/extension.ts (hunk 7 above)
Comment (noklam) [reply src/extension.ts:127]: @jitu5 How can we do this? The `projectMetadata` is only available within the context.

File: bundled/tool/install_telemetry_dependencies.py (hunk 2 above)
Comment (jitu5) [reply bundled/tool/install_telemetry_dependencies.py:24]: @noklam  https://github.com/kedro-org/vscode-kedro/pull/96

File: bundled/tool/check_consent.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+from kedro_telemetry.plugin import _check_for_telemetry_consent
+from kedro_telemetry.plugin import (
+    _get_project_properties,
+    _get_or_create_uuid,
+)
+import sys
+import json
```
Comment (jitu5) [inline bundled/tool/check_consent.py:7]:
```suggestion
from common import update_sys_path
from pathlib import Path
import os
import sys
import json

update_sys_path(
    os.fspath(Path(__file__).parent.parent / "libs"),
    os.getenv("LS_IMPORT_STRATEGY", "useBundled"),
)
```

I tired the above changes and it worked for me.  Let me know if its a right way to do it.

File: kedro-telemetry-requirements.txt [hunk 10]
```diff
@@ -0,0 +1,2 @@
+packaging
+kedro-telemetry>=0.6.0 # First version that does not prompt for telemetry
```
Comment (jitu5) [inline kedro-telemetry-requirements.txt:2]: We need to add `appdirs` as one more dependency for `kedro-telemetry`. As we are not installing all its required dependencies. I got the importError for `appdirs` when I was testing.

Comment (noklam) [reply kedro-telemetry-requirements.txt:2]: Nice found

File: src/extension.ts (hunk 7 above)
Comment (jitu5) [reply src/extension.ts:127]: @noklam 

We can keep it in separate function but need to pass `context` for every call

```
await sendHeapEventWithMetadata(CMD_RESTART_SERVER, context);
```  

```
export async function sendHeapEventWithMetadata(eventName: string, context: ExtensionContext): Promise<void> {
    let projectMetadata: Map<string, any> | undefined;
    let heapUserId: string = '';

    projectMetadata = context.globalState.get(PROJECT_METADATA);

    if (projectMetadata) {
        heapUserId = projectMetadata.get('username');
    }

    sendHeapEvent(eventName, projectMetadata, heapUserId);
}
```

File: src/common/telemetry.ts (hunk 4 above)
Comment (noklam) [reply src/common/telemetry.ts:1]: I guess we can stick with this for now?

File: src/common/telemetry.ts [hunk 11]
```diff
@@ -13,12 +13,21 @@ const HEAP_APPID_PROD = '4039408868'; // todo: Dev server, change it back to pro
 const HEAP_ENDPOINT = 'https://heapanalytics.com/api/track';
 const HEAP_HEADERS = { 'Content-Type': 'application/json' };
 
-export async function sendHeapEvent(eventName: string, properties?: any, identity?: string): Promise<void> {
+export async function sendHeapEvent(commandName: string, properties?: any, identity?: string): Promise<void> {
```
Comment (jitu5) [inline src/common/telemetry.ts:16]: @noklam but we are not passing commandName at all ?

File: bundled/tool/check_consent.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
+from common import update_sys_path
+from pathlib import Path
+import os
+import sys
+import json
+
```
Comment (jitu5) [inline bundled/tool/check_consent.py:6]: I think you accidentally removed 

```
from kedro_telemetry.plugin import _check_for_telemetry_consent
from kedro_telemetry.plugin import (
    _get_project_properties,
    _get_or_create_uuid,
) 
```

File: src/common/telemetry.ts (hunk 11 above)
Comment (noklam) [reply src/common/telemetry.ts:16]: fixed

File: bundled/tool/check_consent.py (hunk 12 above)
Comment (noklam) [reply bundled/tool/check_consent.py:6]: https://github.com/kedro-org/vscode-kedro/pull/84/#discussion_r1750041770

I think this is the reason, I have added it back.

Comment (jitu5) [reply bundled/tool/check_consent.py:13]: @noklam sorry I added the half code as code suggestion. but if I am not wrong, we need to keep `update_sys_path` before we import `kedro_telemetry`

```
from common import update_sys_path
from pathlib import Path
import os
import sys
import json

update_sys_path(
    os.fspath(Path(__file__).parent.parent / "libs"),
    os.getenv("LS_IMPORT_STRATEGY", "useBundled"),
)

from kedro_telemetry.plugin import _check_for_telemetry_consent
from kedro_telemetry.plugin import (
    _get_project_properties,
    _get_or_create_uuid,
)

# Rest of the code
```

**Pull request merged**